src/trace/command.rs

```diff
@@ -0,0 +1,128 @@
+            database_name = event.db.as_str(),
+            command_name = event.command_name.as_str(),
+            request_id = event.request_id,
+            driver_connection_id = event.connection.id,
+            server_connection_id = event.connection.server_id,
+            server_host = event.connection.address.host(),
+            server_port = event.connection.address.port(),
+            service_id = event
+                .service_id
+                .map(|id| id.tracing_representation())
+                .as_deref(),
+            "Command started"
+        );
+    }
+
+    fn handle_command_succeeded_event(&self, event: CommandSucceededEvent) {
+        tracing_debug!(
+            target: COMMAND_TRACING_EVENT_TARGET,
+            client_id: self.client_id.as_ref(),
+            reply = serialize_command_or_reply(event.reply, self.max_document_length_bytes).as_str(),
+            command_name = event.command_name.as_str(),
+            request_id = event.request_id,
+            driver_connection_id = event.connection.id,
+            server_connection_id = event.connection.server_id,
+            server_host = event.connection.address.host(),
+            server_port = event.connection.address.port(),
+            service_id = event
+                .service_id
+                .map(|id| id.tracing_representation())
+                .as_deref(),
+            duration_ms = event.duration.as_millis(),
+            "Command succeeded"
+        );
+    }
+
+    fn handle_command_failed_event(&self, event: CommandFailedEvent) {
+        tracing_debug!(
+            target: COMMAND_TRACING_EVENT_TARGET,
+            client_id: self.client_id.as_ref(),
+            failure = event.failure.tracing_representation(),
+            command_name = event.command_name.as_str(),
+            request_id = event.request_id,
+            driver_connection_id = event.connection.id,
+            server_connection_id = event.connection.server_id,
+            server_host = event.connection.address.host(),
+            server_port = event.connection.address.port(),
+            service_id = event
+                .service_id
+                .map(|id| id.tracing_representation())
+                .as_deref(),
+            duration_ms = event.duration.as_millis(),
+            "Command failed"
+        );
+    }
+}
+
+fn serialize_command_or_reply(doc: bson::Document, max_length_bytes: usize) -> String {
+    let mut ext_json = doc.tracing_representation();
+    truncate_on_char_boundary(&mut ext_json, max_length_bytes);
+    ext_json
+}
+
+// TODO: subject to change based on what exact version of truncation we decide to go with.
+/// Truncates the given string at the closest UTF-8 character boundary >= the provided length.
+/// If the new length is >= the current length, does nothing.
+pub(crate) fn truncate_on_char_boundary(s: &mut String, new_len: usize) {
```

Comment:
Sounds good to me to keep this for now. Can we file a RUST ticket to track the `ceil_char_boundary` portion?

Also, if we did roll our own extJSON serializer that did support graceful truncation, would the spec allow for that? I think it would require a `"truncated": true` field or something to be added to the JSON instead of an ellipsis to ensure the JSON is valid.

Reply:
Sounds good, thanks for another clear + detailed explanation. Java not having any complaints seems like pretty good signal that the "dumb" approach should be good for us :+1: